backend/OKPluginVariableRendering/VariableRendering.cs

```diff
@@ -0,0 +1,326 @@
+
+                            var targetCIAttributes = targetCI.MergedAttributes.Where(a => IsAttributeAllowed(a.Value.Attribute.Name, follow.InputWhitelist, follow.InputBlacklist)).ToList();
+
+                            // NOTE check realtion required trait
+                            if (targetCIAttributes.Where(a => a.Key == follow.RequiredTrait).ToList().Count == 0)
+                            {
+                                continue;
+                            }
+
+                            var allCIAttributes = targetCIAttributes.Select(a => new GatheredAttribute { SourceCIID = targetCI.ID, Name = a.Key, RequiredTrait = follow.RequiredTrait, Value = a.Value.Attribute.Value.ToString() }).ToList();
+
+                            var tmpCIAttributes = new List<GatheredAttribute>();
+                            
+                            foreach (var mapping in follow.AttributeMapping)
+                            {
+                                foreach (var attribute in allCIAttributes)
+                                {
+                                    if (IsAttributeIncludedInSource(attribute.Name, mapping.Source))
+                                    {
+                                        var prio = 0;
+
+                                        switch (predicate[1..])
+                                        {
+                                            case "belongs_to_customer":  prio = customersGroupPrio;
+                                                break;
+                                            case "has_network_interface": prio = networkInterfaceGroupPrio;
+                                                break;
+                                            case "is_attached_to": prio = networkSegmentGroupPrio;
+                                                break;
+                                            case "is_assigned_to": prio = moduleGroupPrio;
+                                                break;
+                                            case "belongs_to_assignment_group": prio = assignmentGroupDefaultPrio;
+                                                break;
+                                            default:
+                                                break;
+                                        }
+
+                                        var a = new GatheredAttribute
+                                        {
+                                            SourceCIID = targetCI.ID,
+                                            Name = GetTargetName(attribute.Name, mapping.Target),
+                                            Value = attribute.Value,
+                                            RequiredTrait = follow.RequiredTrait,
+                                            Priority = prio,
+                                        };
+
+                                        var i = tmpCIAttributes.FindIndex(e => e.Name == attribute.Name);
```

Comment:
Instead of doing this using a List+FindIndex()+RemoveAt(), why not use a Dictionary+TryGetValue()+[] ? Should make the code easier to read AND more performant